tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/Emitter.scala

```diff
@@ -116,9 +117,29 @@ final class Emitter private (semantics: Semantics, outputMode: OutputMode,
   }
 
   private def emitModuleImports(orderedClasses: List[LinkedClass],
-      builder: JSTreeBuilder): Unit = {
+      builder: JSTreeBuilder, logger: Logger): Unit = {
     moduleKind match {
       case ModuleKind.NoModule =>
+        val imports = for {
+          classDef <- orderedClasses
+          JSNativeLoadSpec.Import(module, _) <- classDef.jsNativeLoadSpec
+        } yield {
+          val displayName = decodeClassName(classDef.encodedName)
+          (displayName, module)
+        }
+
+        for {
+          (className, module) <- imports
+        } {
+          logger.error(s"$className needs to be imported from module " +
+              s"'$module' but module support is disabled.")
+        }
```

Comment:
The `Emitter`, especially the non-cached parts, is quite sensitive to the number of allocations (discovered by @nicolasstucki some time ago in a profiling & optimization session). We should make a better effort not to introduce all those tuple allocations and pattern-matching-based `withFilter`s. Consider something like this:
```scala
var importsFound: Boolean = false

for (classDef <- orderedClasses) {
  classDef.jsNativeSpec match {
    case Some(JSNativeLoadSpec.Import(module, _)) =>
      val displayName = decodeClassName(classDef.encodedName)
      logger.error(s"$className needs to be imported from module " +
          s"'$module' but module support is disabled.")
      importsFound = true
    case _ =>
      // ok
}

if (importsFound) {
  sys.error(...)
}
```